bulk_indexer.go

```diff
@@ -257,52 +271,82 @@ func (b *bulkIndexer) Flush(ctx context.Context) (BulkIndexerResponseStat, error
+	if b.maxDocumentRetry != 0 {
+		buf := make([]byte, 1024)
+
+		for k := range b.retryCounts {
+			found := false
+			for _, res := range resp.FailedDocs {
+				if res.Position == k {
+					found = true
+				}
+			}
+			if !found {
+				delete(b.retryCounts, k)
+			}
+		}
 
-	for _, res := range resp.FailedDocs {
-		if res.Status == http.StatusTooManyRequests {
-			startlnIdx := res.Position * 2
-			endlnIdx := startlnIdx + 2
+		tmp := resp.FailedDocs[:0]
+		for _, res := range resp.FailedDocs {
+			if res.Status == http.StatusTooManyRequests {
+				startlnIdx := res.Position * 2
+				endlnIdx := startlnIdx + 2
 
-			if b.gzipw != nil {
-				gr, err := gzip.NewReader(bytes.NewReader(copyBuf))
-				if err != nil {
-					return resp, fmt.Errorf("failed to decompress request payload: %w", err)
+				count := b.retryCounts[res.Position] + 1
+				if count > b.maxDocumentRetry {
+					tmp = append(tmp, res)
+					continue
 				}
-				defer gr.Close()
 
-				seen := 0
+				b.retryCounts[b.itemsAdded] = count
+
+				if b.gzipw != nil {
+					gr, err := gzip.NewReader(bytes.NewReader(b.copyBuf))
+					if err != nil {
+						return resp, fmt.Errorf("failed to decompress request payload: %w", err)
+					}
+					defer gr.Close()
+
+					seen := 0
 
-				n, _ := gr.Read(buf[:cap(buf)])
-				buf = buf[:n]
-				for newlines := bytes.Count(buf, []byte{'\n'}); seen+newlines < startlnIdx; newlines = bytes.Count(buf, []byte{'\n'}) {
-					seen += newlines
 					n, _ := gr.Read(buf[:cap(buf)])
 					buf = buf[:n]
-				}
+					for newlines := bytes.Count(buf, []byte{'\n'}); seen+newlines < startlnIdx; newlines = bytes.Count(buf, []byte{'\n'}) {
+						seen += newlines
+						n, _ := gr.Read(buf[:cap(buf)])
+						buf = buf[:n]
+					}
+
+					start := Indexnth(buf, startlnIdx-seen, '\n') + 1
+					end := Indexnth(buf, endlnIdx-seen, '\n') + 1
 
-				start := Indexnth(buf, startlnIdx-seen, '\n') + 1
-				end := Indexnth(buf, endlnIdx-seen, '\n') + 1
+					for end == 0 {
+						// Add more capacity (let append pick how much).
+						buf = append(buf, 0)[:len(buf)]
 
-				for end == 0 {
-					// Add more capacity (let append pick how much).
-					buf = append(buf, 0)[:len(buf)]
+						n, _ := gr.Read(buf[len(buf):cap(buf)])
+						buf = buf[:len(buf)+n]
 
-					n, _ := gr.Read(buf[len(buf):cap(buf)])
-					buf = buf[:len(buf)+n]
+						end = Indexnth(buf, endlnIdx-seen, '\n') + 1
+					}
 
-					end = Indexnth(buf, endlnIdx-seen, '\n') + 1
+					b.writer.Write(buf[start:end])
+				} else {
+					start := Indexnth(b.copyBuf, startlnIdx, '\n') + 1
+					end := Indexnth(b.copyBuf, endlnIdx, '\n') + 1
+
+					b.writer.Write(b.copyBuf[start:end])
 				}
 
-				b.writer.Write(buf[start:end])
+				resp.RetriedDocs++
+				b.itemsAdded++
 			} else {
-				start := Indexnth(copyBuf, startlnIdx, '\n') + 1
-				end := Indexnth(copyBuf, endlnIdx, '\n') + 1
-
-				b.writer.Write(copyBuf[start:end])
+				tmp = append(tmp, res)
 			}
+		}
 
-			b.itemsAdded++
+		if len(tmp) != 0 {
```

Comment:
Can it be negative?
```suggestion
		if len(tmp) > 0 {
```

Reply:
That was what my comment was about, rather than `!=`, it seems more appropriate to use the `>` operator.

Reply:
+1 on using `>` for better readability when only dealing with positive numbers.